source/common/ext_authz/ext_authz_impl.cc

```diff
@@ -0,0 +1,250 @@
+  using ::envoy::api::v2::auth::AttributeContext_Peer;
+
+  AttributeContext_Peer *peer = new AttributeContext_Peer();
+  ASSERT(peer);
+
+  // Set the address
+  if (local == false) {
+    peer->set_allocated_address(get_pbuf_address(connection->remoteAddress()));
+  } else {
+    peer->set_allocated_address(get_pbuf_address(connection->localAddress()));
+  }
+
+  // Set the principal
+  // Preferably the SAN from the peer's cert or
+  // Subject from the peer's cert.
+  std::string principal;
+  Ssl::Connection* ssl =
+      const_cast<Ssl::Connection*>(connection->ssl());
+  if (ssl != nullptr) {
+    if (local == false) {
+      principal = ssl->uriSanPeerCertificate();
+    
+      if (principal.empty()) {
+        principal = ssl->subjectPeerCertificate();
+      } 
+    } else {
+      principal = ssl->uriSanLocalCertificate();
+    
+      if (principal.empty()) {
+        principal = ssl->subjectLocalCertificate();
+      }
+    }
+    
+  }
+  peer->set_principal(principal);
+
+  if (!service.empty()) {
+    peer->set_service(service);
+  }
+
+  return peer;
+}
+
+::envoy::api::v2::auth::AttributeContext_Peer* CheckRequestGen::get_connection_peer(const Network::Connection& connection, const std::string& service, const bool local) {
+  return get_connection_peer(&connection, service, local);
+}
+
+
+const std::string CheckRequestGen::proto2str(const Envoy::Http::Protocol& p) {
+  using Envoy::Http::Protocol;
+
+  switch(p) {
+    case Protocol::Http10:
+      return std::string("Http1.0");
+    case Protocol::Http11:
+      return std::string("Http1.1");
+    case Protocol::Http2:
+      return std::string("Http2");
+    default:
+      break;
+  }
+  return std::string("unknown");
+}
+
+Envoy::Http::HeaderMap::Iterate CheckRequestGen::fill_http_headers(const Envoy::Http::HeaderEntry &e, void *ctx) {
+  ::google::protobuf::Map< ::std::string, ::std::string >* mhdrs = static_cast<::google::protobuf::Map< ::std::string, ::std::string >*>(ctx);
+  (*mhdrs)[std::string(e.key().c_str(), e.key().size())] = std::string(e.value().c_str(), e.value().size());
+  return Envoy::Http::HeaderMap::Iterate::Continue;
+}
+
+
+::envoy::api::v2::auth::AttributeContext_Request* CheckRequestGen::get_http_request(const Envoy::Http::StreamDecoderFilterCallbacks* callbacks, const Envoy::Http::HeaderMap &headers) {
+  using ::envoy::api::v2::auth::AttributeContext_Request;
+  using ::envoy::api::v2::auth::AttributeContext_HTTPRequest;
+
+  AttributeContext_HTTPRequest *httpreq = new AttributeContext_HTTPRequest();
+  ASSERT(httpreq);
+
+  // Set id
+  // The streamId is not qualified as a const. Although it is as it does not modify the object.
+  Envoy::Http::StreamDecoderFilterCallbacks *sdfc = const_cast<Envoy::Http::StreamDecoderFilterCallbacks *>(callbacks);
+  httpreq->set_id(std::to_string(sdfc->streamId()));
+  
+  #define SET_HDR_IN_HTTPREQ(_hq, _api, mname)                          \
```

Comment:
Is there no other way to do this? Not a huge fan of macro magic (but it's OK if there's really no other way).

Reply:
Let me try to make it a function call.